File: account-server-golang/server_api/READ_RANKING.go
package server_api

import (
	"account-server/database"
	"account-server/ranking"
	"account-server/utils"
	"encoding/json"
	"errors"
	"time"
)

type request_READ_RANKING struct {
	Request_id      string
	Session_id      string
	Ranking_version string
}

func READ_RANKING(message []byte, sendResponse func([]byte)) error {

	var r request_READ_RANKING

	// Comprueba que estan todos los parametros requeridos.
	{
		var m map[string]any
		err := json.Unmarshal(message, &m)
		if err != nil {
			return err
		}
		if !utils.ContainsAllKeys(m, []string{"request_id", "session_id", "ranking_version"}) {
			err = errors.New("READ_RANKING bad input parameters")
			return err
		}

		if _, ok := m["require_test_instance"]; ok { // Evita que las requests, que indican require_test_instance, se ejecuten en una instancia del account server que no es de test.
			if !database.IsTestInstance() {
				return errors.New("require_test_instance")
			}
		}
	}

	err := json.Unmarshal(message, &r)
	if err != nil {
		return err
	}

	var account dict

	// Comprueba si existe una cuenta con esa session_id
	{
		target := []string{"session_id", r.Session_id}
		responseDict, err := database.ReadAccount(target, "session_expire guest_account_id account_name")
		if err != nil {
			return err
		}
		account = responseDict
	}

	responseDict := dict{}

	if account != nil {

		sessionExpired := false

		if account["guest_account_id"] != nil {
			// Para las cuentas guest no se aplica session_expire
		} else {
			// Comprueba que la session no esta caducada
			expireTimeStr := account["session_expire"].(string) // Sera "" si session_expire no esta en account
			if expireTimeStr == "" {
				panic(0)
			}
			expireTime, err := time.Parse(time.RFC3339, expireTimeStr)
			if err != nil {
				panic(0)
			}
			sessionExpired = expireTime.Before(time.Now())
		}

		if !sessionExpired {

			var account_id string
			if v, ok := account["account_name"]; ok {
				account_id = v.(string)
			} else if v, ok := account["guest_account_id"]; ok {
				account_id = v.(string)
			} else {
				panic(0)
			}

			var accountPositions []any
			content := ""

			cr_version, cr_content, cr_account_positions := ranking.GetCurrentRanking(account_id)
			if r.Ranking_version != cr_version {
				content = cr_content
				accountPositions = cr_account_positions
			}

			responseDict["session_id"] = r.Session_id
			responseDict["ranking"] = dict{
				"version":           cr_version,
				"content":           content,
				"account_positions": accountPositions,
			}

		} else {
			responseDict["error"] = "Session is expired."
		}

	} else {
		responseDict["error"] = "No account found with this session_id."
	}

	// Envia la respuesta

	responseDict["request_id"] = r.Request_id

	response, err := json.Marshal(responseDict)
	if err != nil {
		return err
	}
	sendResponse(response)

	return nil
}

File: account-server/server_api/READ_RANKING.py
from datetime import datetime
from utils import str_to_date

# READ_RANKING
# 1. El cliente envia READ_RANKING(session_id) al servidor.
# 2. El servidor comprueba que existe esa session_id y no esta caducada.
# 3. El servidor responde con el ranking.
async def READ_RANKING(req, ws, send_message, accounts_cache, ranking):

    # Aqui se debe comprobar que la req es correcta y en caso contrario se ignora sin enviar nada al cliente.
    if not all(k in req for k in ('request_id', 'session_id', 'ranking_version')):
        print("ERROR: READ_RANKING bad input parameters.")
        return

    error_response = {
        "error": "Invalid ticket.",
        "request_id": req['request_id']
    }  

    valid = False

    ### Comprueba si existe esa session_id

    session_id = req["session_id"]

    query = {"session_id": session_id}
    projection = 'session_expire guest_account_id account_name'
    response = await accounts_cache.read_account(req, ws, send_message, query, projection)
    if response is None:
        return

    ### Comprueba si la session esta caducada

    if 'session_expire' in response:
        if 'guest_account_id' in response or str_to_date(response['session_expire']) >= datetime.now():

            ranking.update_ranking() # Actualiza ranking

            account_name = response['account_name']
            account_positions = ranking.get_account_positions(account_name)

            del response['session_expire']
            response['session_id'] = session_id
            response["ranking"] = {
                'version': ranking.ranking_version,
                'content': ranking.json_serializable_ranking if ranking.ranking_version != req['ranking_version'] else '',
                'account_positions': account_positions
            }
            response["request_id"] = req['request_id']

            await send_message(response, ws)  # send response
            valid = True

        else:
            error_response["error"] = "Session is expired."
    else:
        error_response["error"] = "No account found."

    ### Envia error al cliente.

    if not valid:
        await send_message(error_response, ws)

    

    

